src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,387 @@
+          auto builder = callContext.initResults(
+              capnp::MessageSize { serialized.size() / 8 + 8, 0 }).initException();
+          builder.setV8Serialized(kj::mv(serialized));
+          co_return;
+        }
+
+        auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+
+        // We have a function, so let's call it and serialize the result for RPC.
+        // We account for both the JS method throwing and our serialization attempt failing.
+        // If the function returns a promise we will wait for the promise to finish so we can
+        // serialize the result.
+        auto result = co_await callFuncAndSerialize(js, fn, serializedArgs);
+
+        auto builder = callContext.initResults(capnp::MessageSize { result.data.size() / 8 + 8, 0 });
+        if (result.caughtException) {
+          builder.initException().setV8Serialized(kj::mv(result.data));
+        } else {
+          builder.initValue().setV8Serialized(kj::mv(result.data));
         }
-    ));
+      }));
+    } catch(kj::Exception& e) {
+      if (auto desc = e.getDescription();
+          !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
+        LOG_EXCEPTION("JsRpcTargetCall"_kj, e);
+      }
+    }
+    // Upon returning, we want to fulfill the callPromise so customEvent can continue executing.
+    KJ_DEFER(callFulfiller->fulfill(););
+    co_return;
   }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // The following names are registered by the Workers Runtime and cannot be called over RPC.
+  bool isRegisteredName(kj::StringPtr name) {
+    // TODO(now): block all methods inherited from Object.
+    if (name == "fetch"_kjc ||
+        name == "connect"_kjc ||
+        name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  // If the `methodName` is a known public method, we'll return it.
+  inline kj::OneOf<v8::Local<v8::Function>, kj::String> tryGetFn(
+      Worker::Lock& lock,
+      kj::Maybe<kj::StringPtr> entrypointName,
+      IoContext& ctx,
+      kj::StringPtr methodName) {
+
+    auto& handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, ctx.getActor()),
+        "Failed to get handler to worker.");
+
+    if (isRegisteredName(methodName)) {
+      return kj::str("'", methodName, "' is a registered method and cannot be called over RPC."_kj);
+    }
+    auto handle = handler.self.getHandle(lock);
+    auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+    auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+    if (!(fnHandle->IsFunction() && !fnHandle->IsPrivate())) {
+      // We can't return this because it's not a function, or if it is, it's private.
+      return kj::str("The RPC receiver does not implement the requested method."_kj);
+    }
+    return fnHandle.As<v8::Function>();
```

Comment:
Probably would be better to wrap this in a `jsg::Function<jsg::Promise<jsg::Value>(...)>` if possible.